### DISCUSSION TITLE
macOS Ventura?

### DISCUSSION THREAD

**Discussion by infoXczar** [ExistentialAudio/BlackHole#674] [Q&A] [answered]
I just updated to macOS Ventura and now my multi-output devices are crackling, stuttering, and possibly also causing video drag and other issues. I reinstalled the latest Blackhole just to be sure I was running the latest release, and tried making a new multi-output device. As soon as I switch the Mac's output to it, I start noticing the crackling and stuttering. It goes away as soon as I switch back to e.g. MacBook Pro Speakers.

Anyone else having these issues? Any suggestions on what I can try to ameliorate the problems? I'm literally just doing MacBook Pro Speakers as Primary Device and BlackHole 2ch w/Drift Correction as the only other device, for the purposes of capturing system audio and routing into OBS, so it's nothing fancy...

**Comment by devinroth** [comment]
I tested on the beta and everything worked normally. I’ll install the release once I have a chance and let you know if I’m able to replicate your issues.

Devin Roth

**Reply by infoXczar** [reply]
Interestingly, today the problem didn't recur, though I didn't do anything different that I know of... I have a very specific ordered checklist of things I do after restarting before streaming, and today everything seemed fine. If I notice a recurrence, I'll try to let you know more information. Thanks!

**Reply by DraganBocevski** [reply]
I had the exact same issue. OS restart didn't help. Audio stutters and video drags. 
I think the problem started appearing after upgrading Blackhole to latest release on Ventura. The old version (0.2.x) worked fine even on Ventura.

It appears to be solved after reinstalling Blackhole (might help somebody):
brew reinstall --cask blackhole-2ch
brew reinstall --cask blackhole-16ch

**Comment by devinroth** [comment, marked as answer]
Just tested in Ventura and multi-outputs with BlackHole are working normally for me.

**Reply by infoXczar** [reply]
Thank you for looking into it! Must have been a one-time thing.

**Comment by MartGHub** [comment]
As far as I know I'm running the latest version (I downloaded it again last week). Since upgrading MacOS to Ventura I'm getting very high levels of latency when I try to play a YT video (e.g. a backing track or song I am learning) and use Helix Native (a guitar/bass plug-in running in GarageBand) at the same time. I'm getting the same latency issue with other sound sources also, such as Ireal Pro (which isn't streaming information from the internet). When I switch my audio routing to just use my audio interface and take Blackhole out of the signal path the latency issues disappear.

The issue is recurring after multiple days/system restarts.

Any advice on how to address this?  Would be very grateful.

**Comment by DaveHamilton** [comment]
I'm still in the "final" stages of isolating my problem, but I routinely use an aggregate device of my Thunderbolt Interface (Quantum 2626) and Blackhole-16ch, and as soon as I upgraded to Ventura last week, I started getting 1-second-long cutouts in my audio every couple of minutes, each followed by a 2-second-long coreaudiod CPU spike (like the buffer was getting too full, coreaudiod was punting, and then playing catchup as it came back online). 

This morning, on a hunch, I disabled BlackHole from that aggregate device and replaced it with a couple of Loopback (paid, from Rogue Amoeba) devices, and cannot (yet?) get the problem to recur.

Like I said, I'm still in the "testing" phase of this theory, but thus far it seems to be playing out and I figured I'd share. I know the Rogue Amoeba folks had to rewrite quite a bit of their Audio Capture Engine (ACE) for Ventura, and so perhaps there's a similar issue that needs to be addressed before BlackHole is reliable with Apple's newly-released macOS?

**Reply by MartGHub** [reply]
In case helpful to know I have fixed the issue I was having with Blackhole 2 ch.  It was as simple as having to uninstall the existing version of Blackhole 2 ch I was running and then downloading and re-installing the new version.  I had assumed that when I downloaded and installed the new version it would overwrite the existing version but that doesn't seem to have happened (though I confess I wasn't able to verify that).  

Best of luck getting your issue sorted.

M

**Reply by DaveHamilton** [reply]
I think @devinroth's supposition that this is an issue with aggregate audio devices in Ventura has some merit: In a three-hour recording session this morning, I had this issue happen twice, whereas last week it was *far* more regular. So while disabling Blackhole in favor of Loopback definitely made things _remarkably_ better, it's not completely solved. Seems like whatever blackhole is/was doing exacerbated this issue, but the issue itself still remains. 

Each of the two times that I had this issue today, I very clearly "caused" it by launching a new app or opening a web browser window (aka, I did _something_ that messed with the CPU enough that audio timing was disrupted and I wound up with one of these hiccups). Prior to removing blackhole from my aggregate device, though, I had these hiccups ever few minutes without any obvious trigger action on my part.

Hopefully that helps diagnose?

**Comment by devinroth** [comment]
I think this is more an issue with Aggregates than BlackHole. Just to confirm drift correction is enabled on BlackHole right?

Devin Roth

**Reply by devinroth** [reply]
Check to make sure you have the latest. In BlackHole.driver/contents/info.plist will show a version number.

**Reply by DaveHamilton** [reply]
Thanks, @devinroth — Drift correction is enabled on BlackHole, yep! 

My version string is: 

```
<key>CFBundleShortVersionString</key>
 <string>0.4.0</string>

```
I installed with homebrew, which I think keeps it up-to-date (but please correct me if I'm wrong on that).

Thanks!

**Reply by DaveHamilton** [reply]
Indeed, `brew list blackhole-16ch` shows: BlackHole16ch.v0.4.0.pkg (Pkg), and `brew upgrade blackhole-16ch` says "Warning: Not upgrading blackhole-16ch, the latest version is already installed".

**Comment by devinroth** [comment]
I’m pretty sure it’s an issue with Aggregate devices. Are you willing to do some tests for me?

And maybe test some software I’m working on?

Devin Roth

**Reply by DaveHamilton** [reply]
Happy to test, yep!

**Comment by devinroth** [comment]
I suspect the reason BlackHole exacerbates the issue is due to the lower latency vs Loopback. There is less room for error with BlackHole.

I’ll get some stuff together for you to test hopefully today. Just got back from Christmas vacation.

Devin Roth

**Reply by DaveHamilton** [reply]
Makes sense. The reason I went with blackhole in the beginning was because my workflow back then relied upon a virtual driver for me to monitor my own signal, and loopback was (way) too slow then. It's gotten better, but my workflow has also changed in a way that's not dependent on that.

Still, happy to test, and would love to get to the bottom of this. Thanks!

**Reply by devinroth** [reply]
I’m curious. What’s your current workflow?

Okay test this out. Run AudioPlayThroughTest with Terminal. 
```
% AudioPlayThrough inputDeviceUID outputDeviceUID
```
[AudioPlayThrough.zip](https://github.com/ExistentialAudio/BlackHole/files/10327097/AudioPlayThrough.zip)

**Reply by devinroth** [reply]
You can get the UID from the CoreAudioDiagnostics app I wrote. You may need to change permissions to run these programs. 
```
% chmod 755 
```

**Reply by devinroth** [reply]
I wrote this a few years ago in response to issues with Aggregates and developed a better audio drift correction algorithm. I've included this in a number of projects that I've worked on. AudioPlayThrough should be really low latency. Could be even lower with smaller buffer sizes. 

I suspect that Apple broke Aggregates / Multi-Output devices in Ventura. I've been working on an app that will solve this problem. Hope to release in the near future. What I posted here is the barebones.

**Reply by devinroth** [reply]
I think in your case you should set this up by outputting to BlackHole. Recording from BlackHole. And monitoring audio with `AudioPlayThrough BlackHole2ch_UID OutputSpeakersUID `.

**Reply by devinroth** [reply]
@DaveHamilton were you able to take this for a spin yet?
